src/API/FileController.ts

```diff
@@ -0,0 +1,95 @@
+import * as FormData from 'form-data'
+import * as fs from 'fs'
+import { inject, injectable } from 'inversify'
+import { Collection, Db } from 'mongodb'
+import fetch, {Response} from 'node-fetch'
+import * as Pino from 'pino'
+import { map } from 'ramda'
+
+import { childWithFileName } from 'Helpers/Logging'
+import { minutesToMiliseconds  } from 'Helpers/Time'
+import { asyncPipe } from 'Helpers/asyncPipe'
+
+interface IPFSFileResponseJson {
+  readonly Hash: string
+}
+
+interface FileResponseJson {
+  readonly hash: string
+}
+
+interface File {
+  readonly path: string
+}
+
+enum LogTypes {
+  info = 'info',
+  trace = 'trace',
+  error = 'error',
+}
+
+export interface FileControllerConfiguration {
+  readonly ipfsUrl: string
+}
+
+export const convertJson = (x: IPFSFileResponseJson) => ({
+  hash: x.Hash,
+})
+
+@injectable()
+export class FileController {
+  private readonly logger: Pino.Logger
+  private readonly db: Db
+  private readonly collection: Collection
+  private readonly ipfsUrl: string
+
+  /* tslint:disable-next-line:max-line-length */
+  constructor(@inject('Logger') logger: Pino.Logger, @inject('DB') db: Db, @inject('FileControllerConfiguration') configuration: FileControllerConfiguration) {
+    this.logger = childWithFileName(logger, __filename)
+    this.db = db
+    this.collection = this.db.collection('files')
+    this.ipfsUrl = configuration.ipfsUrl
+  }
+
+  private log = (logType: LogTypes) => (label: string) => async <T extends any>(value?: T ) => {
+    this.logger[logType]({ value }, label)
+    return value
+  }
+
+  private uploadStream = (stream: fs.ReadStream) => {
+    const formData = new FormData()
+    formData.append('file', stream)
+    return fetch(`${this.ipfsUrl}/api/v0/add`, {
+      method: 'post',
+      body: formData,
+      timeout: minutesToMiliseconds(10),
+    })
+  }
+
+  private getResponseJson = (x: Response) => x.json()
+
+  private addFileToIPFS = (fileStream: fs.ReadStream) => this.uploadStream(fileStream)
+
+  private storeIPFSHash = async (response: FileResponseJson) => {
+    const { hash } = response
+    await this.collection.insertOne({ hash })
+    return response
+  }
+
+  private handleFile = asyncPipe(
```

Comment:
Not sure we need all these log traces in this async pipe. Logs in the main might be sufficient.

Reply:
The strategy for now is too much is better than too little, regarding logs. We can downsize after we're in production and know everything's going well.